chia/data_layer/data_layer.py

```diff
@@ -136,3 +141,137 @@ async def get_root(self, store_id: bytes32) -> Optional[bytes32]:
             self.log.error(f"Failed to get root for {store_id.hex()}")
             return None
         return res.node_hash
+
+    async def _validate_batch(
+        self,
+        tree_id: bytes32,
+        to_check: List[SingletonRecord],
+        min_generation: int,
+        max_generation: int,
+    ) -> bool:
+        last_checked_hash: Optional[bytes32] = None
+        for record in to_check:
+            # Ignore two consecutive identical root hashes, as we've already validated it.
+            if last_checked_hash is not None and record.root == last_checked_hash:
+                continue
+            # Pick the latest root in our data store with the desired hash, before our already validated data.
+            root: Optional[Root] = await self.data_store.get_last_tree_root_by_hash(
+                tree_id, record.root, max_generation
+            )
+            if root is None or root.generation < min_generation:
+                return False
+
+            self.log.info(f"Validated chain hash {record.root} in downloaded datastore.")
+            max_generation = root.generation
+            last_checked_hash = record.root
+
+        return True
+
+    async def fetch_and_validate(self, subscription: Subscription) -> None:
+        tree_id = subscription.tree_id
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id)
+        if singleton_record is None:
+            return
+        if singleton_record.generation == uint32(0):
+            return
+        old_root: Optional[Root] = None
+        if await self.data_store.tree_id_exists(tree_id=tree_id):
+            old_root = await self.data_store.get_tree_root(tree_id=tree_id)
```

Comment:
if you try/catch on get_tree_root(tree_id=tree_id) you can remove tree_id_exists and call the db only once